fish-rust/src/common.rs

```diff
@@ -1,13 +1,1714 @@
+        None => panic!("not a valid char"),
+    }
+}
+
+/// A user-visible job ID.
+pub type JobId = i32;
+
+/// The non user-visible, never-recycled job ID.
+/// Every job has a unique positive value for this.
+pub type InternalJobId = u64;
+
+/// Exits without invoking destructors (via _exit), useful for code after fork.
+fn exit_without_destructors(code: i32) -> ! {
+    unsafe {
+        libc::_exit(code);
+    }
+}
+
+/// Save the shell mode on startup so we can restore them on exit.
+static SHELL_MODES: Lazy<Mutex<libc::termios>> = Lazy::new(|| Mutex::new(unsafe { mem::zeroed() }));
+
+/// The character to use where the text has been truncated. Is an ellipsis on unicode system and a $
+/// on other systems.
+pub fn get_ellipsis_char() -> char {
+    char::from_u32(ELLIPSIS_CHAR.load(Ordering::Relaxed)).unwrap()
+}
+
+static ELLIPSIS_CHAR: AtomicU32 = AtomicU32::new(0);
+
+/// The character or string to use where text has been truncated (ellipsis if possible, otherwise
+/// ...)
+pub static mut ELLIPSIS_STRING: Lazy<&'static wstr> = Lazy::new(|| L!(""));
+
+/// Character representing an omitted newline at the end of text.
+pub fn get_omitted_newline_str() -> &'static wstr {
+    unsafe { &OMITTED_NEWLINE_STR }
+}
+
+static mut OMITTED_NEWLINE_STR: Lazy<&'static wstr> = Lazy::new(|| L!(""));
+
+pub fn get_omitted_newline_width() -> usize {
+    unsafe { OMITTED_NEWLINE_STR.len() }
+}
+
+static OBFUSCATION_READ_CHAR: AtomicU32 = AtomicU32::new(0);
+
+pub fn get_obfuscation_read_char() -> char {
+    char::from_u32(OBFUSCATION_READ_CHAR.load(Ordering::Relaxed)).unwrap()
+}
+
+/// Profiling flag. True if commands should be profiled.
+pub static G_PROFILING_ACTIVE: RelaxedAtomicBool = RelaxedAtomicBool::new(false);
+
+/// Name of the current program. Should be set at startup. Used by the debug function.
+pub static mut PROGRAM_NAME: Lazy<&'static wstr> = Lazy::new(|| L!(""));
+
+#[cfg(windows)]
+/// Set to false if it's been determined we can't trust the last modified timestamp on the tty.
+pub const HAS_WORKING_TTY_TIMESTAMPS: bool = false;
+#[cfg(not(windows))]
+/// Set to false if it's been determined we can't trust the last modified timestamp on the tty.
+pub const HAS_WORKING_TTY_TIMESTAMPS: bool = true;
+
+/// A global, empty string. This is useful for functions which wish to return a reference to an
+/// empty string.
+pub static G_EMPTY_STRING: WString = WString::new();
+
+/// A global, empty wcstring_list_t. This is useful for functions which wish to return a reference
+/// to an empty string.
+pub static G_EMPTY_STRING_LIST: Vec<WString> = vec![];
+
+/// A function type to check for cancellation.
+/// \return true if execution should cancel.
+pub type CancelChecker = dyn Fn() -> bool;
+
+/// Converts the narrow character string \c in into its wide equivalent, and return it.
+///
+/// The string may contain embedded nulls.
+///
+/// This function encodes illegal character sequences in a reversible way using the private use
+/// area.
```

Comment:
The documentation of str2wcstring says

	/// Converts the narrow character string \c in into its wide equivalent, and return it.
	///
	/// This function encodes illegal character sequences in a reversible way using the private use
	/// area.
	pub fn str2wcstring(inp: &[u8]) -> WString;

Surprisingly, it may produce surrogates (0xD800..0xDFFF) and characters that are outside the Unicode codespace (> 0x10FFFF) because mbrtowc does not prevent this.

For instance

	#include <assert.h>
	#include <locale.h>
	#include <stdint.h>
	#include <stdio.h>
	#include <string.h>
	#include <wchar.h>

	int main() {
	    setlocale(LC_ALL, "");
	    wchar_t wc = L'\0';
	    uint8_t input[4] = {245, 155, 155, 143};
	    mbstate_t ps;
	    memset(&ps, '\0', sizeof(mbstate_t));
	    ssize_t ret = mbrtowc(&wc, (const char*)&input[0], 4, &ps);
	    assert(ret == 4);
	    assert(wc == 0x15b6cf);
	}

Rusts `char::from_u32` fails for such code points, breaking many assumptions.
I think due to the broken invariant of `WString`, its `char_at()` does not fail but silently return an invalid character.

To make this easier to handle, I'll encode invalid code points with the private-use-are trick.

Reply:
Yuck. Yeah your proposed private-use-area trick seems right to me.